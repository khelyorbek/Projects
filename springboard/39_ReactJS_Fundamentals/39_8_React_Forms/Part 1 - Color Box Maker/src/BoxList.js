import React, {useState} from 'react';
import NewBoxForm from './NewBoxForm';
import Box from './Box';

const BoxList = () => {
    const INITIAL_STATE = [
        { width: '300', height: '300', bgColor: 'DeepSkyBlue'}
    ]

    // Place your state that contains all of the boxes here.
    const [allBoxes, setAllBoxes] = useState(INITIAL_STATE);

    // custom method to handle additions to our allBoxes state
    const addBox = (newBox) => {
        setAllBoxes(items => [...allBoxes, newBox]);
    }

    // custom method to handle removal of an item from the allboxes state
    const removeBox = (colorToRemove) => {
        setAllBoxes(allBoxes.filter(obj => obj.bgColor !== colorToRemove));
    }

    // This component should render all of the Box components along with the NewBoxForm component
    return (
    <>
    <NewBoxForm addBox={addBox}/>
    
    <div>
        {allBoxes.map(({width, height, bgColor}) => <Box width={width} height={height} bgColor={bgColor} key={bgColor} removeBox={removeBox} />)}
    </div>
    
    </>
        )
}

export default BoxList;