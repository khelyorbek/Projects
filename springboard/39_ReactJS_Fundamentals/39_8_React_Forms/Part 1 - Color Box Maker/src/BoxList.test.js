import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom'
import BoxList from './BoxList';

// smoke test
it('renders without crashing', function() {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("renders the form properly", function() {
  const { getByText } = render(<BoxList />);
  const widthInput = getByText("Width:");
  const heightInput = getByText("Height:");
  const submitBtn = getByText("Create new box");

  expect(widthInput).toBeInTheDocument();
  expect(heightInput).toBeInTheDocument();
  expect(submitBtn).toBeInTheDocument();
});

it("clears inputs after typing", function() {
  const { getByText, getByPlaceholderText } = render(<BoxList />);
  const widthInput = getByPlaceholderText("Type in the width here");
  const heightInput = getByPlaceholderText("Type in the height here");
  const submitBtn = getByText("Create new box");

  fireEvent.change(widthInput, {target: {value: '300'}});
  fireEvent.change(heightInput, {target: {value: '300'}});

  // click on the button
  fireEvent.click(submitBtn);

  // is there a difference?
  expect(widthInput).not.toHaveTextContent('300');
  expect(heightInput).not.toHaveTextContent('300');
})

it("creates a new box", function() {
  const { getByText, getByPlaceholderText, getAllByText } = render(<BoxList />);
  const widthInput = getByPlaceholderText("Type in the width here");
  const heightInput = getByPlaceholderText("Type in the height here");
  const submitBtn = getByText("Create new box");

  expect(getAllByText("REMOVE").length).toEqual(1);

  fireEvent.change(widthInput, {target: {value: '300'}});
  fireEvent.change(heightInput, {target: {value: '300'}});

  // click on the button
  fireEvent.click(submitBtn);

  // is there a difference?
  expect(getAllByText("REMOVE").length).toEqual(2);
})

it("removes an existing box", function() {
  const { getByText, getByPlaceholderText, getAllByText } = render(<BoxList />);
  const widthInput = getByPlaceholderText("Type in the width here");
  const heightInput = getByPlaceholderText("Type in the height here");
  const submitBtn = getByText("Create new box");

  fireEvent.change(widthInput, {target: {value: '300'}});
  fireEvent.change(heightInput, {target: {value: '300'}});

  // click on the button
  fireEvent.click(submitBtn);

  // is there a difference?
  expect(getAllByText("REMOVE").length).toEqual(2);

  fireEvent.click(getAllByText("REMOVE")[0]);

  expect(getAllByText("REMOVE").length).toEqual(1);
})
