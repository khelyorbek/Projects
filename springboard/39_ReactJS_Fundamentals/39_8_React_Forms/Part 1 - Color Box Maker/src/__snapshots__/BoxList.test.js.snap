// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      placeholder="Type in the width here"
      type="number"
      value=""
    />
    <br />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      placeholder="Type in the height here"
      type="number"
      value=""
    />
    <br />
    <label
      for="bgColor"
    >
      Background Color:
    </label>
    <input
      id="bgColor"
      name="bgColor"
      type="color"
      value="#000000"
    />
    <br />
    <button>
      Create new box
    </button>
  </form>
  <div>
    <div
      style="display: flex; font-size: 3rem; align-items: center; justify-content: center; text-align: center; width: 300px; height: 300px; background-color: DeepSkyBlue;"
    >
      REMOVE
    </div>
  </div>
</DocumentFragment>
`;
